// @solid-primitives/audio/dist/index.cjs
'use strict';

var solidJs = require('solid-js');
var web = require('solid-js/web');
var utils = require('@solid-primitives/utils');
var staticStore = require('@solid-primitives/static-store');

// src/index.ts
exports.AudioState = /* @__PURE__ */ ((AudioState2) => {
  AudioState2["LOADING"] = "loading";
  AudioState2["PLAYING"] = "playing";
  AudioState2["PAUSED"] = "paused";
  AudioState2["COMPLETE"] = "complete";
  AudioState2["STOPPED"] = "stopped";
  AudioState2["READY"] = "ready";
  AudioState2["ERROR"] = "error";
  return AudioState2;
})(exports.AudioState || {});
var unwrapSource = (src) => {
  let player;
  if (src instanceof HTMLAudioElement) {
    player = src;
  } else {
    player = new Audio();
    player[typeof src === "string" ? "src" : "srcObject"] = src;
  }
  return player;
};
exports.makeAudio = (src, handlers = {}) => {
  if (web.isServer) {
    return {};
  }
  const player = unwrapSource(src);
  const listeners = (enabled) => {
    Object.entries(handlers).forEach(
      ([evt, handler]) => player[enabled ? "addEventListener" : "removeEventListener"](
        evt,
        handler
      )
    );
  };
  solidJs.onMount(() => listeners(true));
  solidJs.onCleanup(() => {
    player.pause();
    listeners(false);
  });
  return player;
};
exports.makeAudioPlayer = (src, handlers = {}) => {
  if (web.isServer) {
    return {
      pause: utils.noop,
      play: async () => utils.noop(),
      player: {},
      seek: utils.noop,
      setVolume: utils.noop
    };
  }
  const player = exports.makeAudio(src, handlers);
  const play = () => player.play();
  const pause = () => player.pause();
  const seek = (time) => (
    // eslint-disable-next-line @typescript-eslint/no-unnecessary-condition
    player.fastSeek ? player.fastSeek(time) : player.currentTime = time
  );
  const setVolume = (volume) => player.volume = volume;
  return { play, pause, seek, setVolume, player };
};
exports.createAudio = (src, playing, volume) => {
  if (web.isServer) {
    return [
      {
        state: "loading" /* LOADING */,
        currentTime: 0,
        duration: 0,
        volume: 0,
        player: {}
      },
      {
        seek: utils.noop,
        setVolume: utils.noop,
        play: async () => utils.noop(),
        pause: utils.noop
      }
    ];
  }
  const player = unwrapSource(utils.access(src));
  const [store, setStore] = staticStore.createStaticStore({
    state: "loading" /* LOADING */,
    player,
    currentTime: 0,
    get duration() {
      return this.player.duration;
    },
    get volume() {
      return this.player.volume;
    }
  });
  const { play, pause, setVolume, seek } = exports.makeAudioPlayer(store.player, {
    loadeddata: () => {
      setStore({
        state: "ready" /* READY */,
        duration: player.duration
      });
      if (playing && playing()) {
        play().catch((e) => {
          if (e.name === "NotAllowedError") {
            setStore("state", "error" /* ERROR */);
          }
        });
      }
    },
    timeupdate: () => setStore("currentTime", player.currentTime),
    loadstart: () => setStore("state", "loading" /* LOADING */),
    playing: () => setStore("state", "playing" /* PLAYING */),
    pause: () => setStore("state", "paused" /* PAUSED */),
    error: () => setStore("state", "error" /* ERROR */)
  });
  if (src instanceof Function) {
    solidJs.createEffect(() => {
      const newSrc = utils.access(src);
      if (newSrc instanceof HTMLAudioElement) {
        setStore("player", () => newSrc);
      } else {
        store.player[typeof newSrc === "string" ? "src" : "srcObject"] = newSrc;
      }
      seek(0);
    });
  }
  if (playing) {
    solidJs.createEffect(() => playing() ? play() : pause());
  }
  if (volume) {
    solidJs.createEffect(() => setVolume(volume()));
    setVolume(volume());
  }
  return [store, { seek, play, pause, setVolume }];
};
